### ISSUE TITLE
Webpack multi-compiler mode with 21 configs causes OOM

### ISSUE THREAD

**Issue by ayc92** [webpack/webpack#2440]
The app I'm working on is made up of 4 pages, each needs its own bundle. In addition, I need to localize my code to 21 languages, so I've decided to generate 4 bundles per language. I'm running into an OOM memory exception, but I'm not sure whether this is due to the number of bundles I'm generating, or a problem with something in my configs. I'm running webpack in a docker container (via docker-compose). The docker container is allowed to use as much memory as the docker machine has. Mine has 4096MB.

I've tried increasing the node process memory as well as threadpool size, both to no avail. I've also made sure that I'm not using the uglify plugin on my development env, and i've also tried disabling sourcemaps. Nothing has worked.

Here's part of my `webpack.config.js` file.

```
function getPlugins(language) {
    var validationLanguage = validationLanguagesMap[language];
    var plugins = {
        default: [
            new webpack.DefinePlugin({
                'process.env.NODE_ENV': isProdEnv() ? '"production"' : '"development"',
                'process.env.LOCALE': JSON.stringify(language),
                'process.env.JQUERY_VALIDATION_LOCALE': JSON.stringify(validationLanguage),
            }),
            new webpack.optimize.MinChunkSizePlugin({minChunkSize: 10000}),
            // Only use the "en" (default) locale for moment
            new webpack.ContextReplacementPlugin(/moment[\/\\]lang/, /^\.\/(NO_MATCH)$/),
            new webpack.ProvidePlugin({
                _: 'underscore',
                $: 'jquery',
                JQuery: 'jquery',
            }),
            new webpack.IgnorePlugin(/jquery-validation\/dist\/localization\/messages_en-us/),
            new webpack.IgnorePlugin(/jquery-ui-datepicker\/messages-en-us/),
            new BundleTracker({filename: 'dist/webpack-stats.json'}),
        ],

        production: [
            // Optimization
            new webpack.optimize.OccurenceOrderPlugin(),
            new webpack.optimize.DedupePlugin(),
            new webpack.optimize.UglifyJsPlugin(),
        ],

        development: []
    };

    var env = isProdEnv() ? 'production' : 'development';
    return plugins.default.concat(plugins[env]);
}

module.exports = bundleLanguages.map(function (language) {
    var entries = {
        book: ['book/main'],
        floor: ['floor/main'],
        grid: ['grid/main'],
        mgmt: ['mgmt/main'],
        consumer: ['consumer/main'],
    };
    entries['microsite-' + language] = ['microsite/main'];

    return {
        context: __dirname,
        devtool: enableSourceMaps(),

        entry: entries,

        output: {
            path: path.resolve('./dist/bundles/'),
            filename: '[name]-[hash].js'
        },

        plugins: getPlugins(language),

        watchOptions: {
            // Docker doesn't support inotify, so use polling instead
            poll: true
        },

        // Hack around modules with dirty modules syntax
        module: {
            loaders: [
                {
                    test: /\.jsx?$/,
                    // Don't parse vendors libs with babel
                    exclude: [
                        /node_modules/,
                        /www\/js\/libs/,
                        /www\/js\/locale/,
                    ],
                    loader: 'babel',
                    cacheDirectory: true,
                },
                {
                    test: /\.tpl$/,
                    loader: 'underscore-template-loader',
                    query: {
                        prependFilenameComment: __dirname,

                        // Escape expressions like:  {{ myVariable }}
                        // Note that we ensure this will not match {{= var }}, since this pattern is
                        // checked before "interpolate"
                        escape: '\\{\\{(?!\\=)(.+?)\\}\\}',

                        // Interpolate expressions like:  {{= myVariable }}
                        interpolate: '\\{\\{=(.+?)\\}\\}',

                        // Evaluate expressions like:  {% if (myVariable) { %}
                        evaluate: '\\{\\%(.+?)\\%\\}'
                    }
                },
                {
                    test: /\.json$/,
                    loader: 'json-loader'
                },
                {
                    test: /\.(jpg|png)$/,
                    loader: 'file-loader'
                },
                {
                    test: /\.svg$/,
                    loaders: [
                        'svg-sprite',
                        'svgo-loader?' + JSON.stringify({
                            plugins: [
                                {removeTitle: true},
                                // We want to remove the fill attributes to allow
                                // coloring via CSS
                                {removeAttrs: {attrs: ['fill']}}
                            ]
                        })
                    ]
                },
                {
                    test: path.resolve('www/js/libs/jquery-1.7.2.min'),
                    loader: 'exports?window.jQuery'
                },
                {
                    test: path.resolve('www/js/libs/libphonenumber'),
                    loader: 'exports?i18n'
                },
                {
                    test: require.resolve('backbone'),
                    loader: 'exports?window.Backbone'
                }
            ],
            noParse: _.values(bundleDepsMap),
        },

        resolve: {
            root: path.resolve('www/js/modules'),
            modulesDirectories: ['node_modules', 'consumer', 'utils'],
            extensions: ['', '.js', '.jsx'],
            alias: _.extend(addJQueryUIDatepickerLanguage(language, bundleDepsMap), {
                'braintree': path.resolve('www/js/libs/braintree-1.3.10.js'),
            }),
        },
    };
});
```
If I'm using a loop to generate the configs, does webpack somehow keep all of the generate chunks in memory before writing them out?

**Comment by bebraw** [comment]
Does it work if you process only the first item of the array? If so, it's definitely related to the array somehow.

**Comment by ayc92** [comment]
Yes, it works if I process just the first item. I'm trying to figure out the minimum set that causes an OOM error. I will get back to you on that.

**Comment by ayc92** [comment]
@bebraw: I dug into the source code a bit, and I'm wondering if it's possible that there's memory leak in the multi-compiler code path, but that it's only exacerbated by my situation, since I generate a large number of configs.

The code I'm looking at is in `Compiler.js::Watching.prototype._go`. The `_go` function is called for every `Watching` instance, and one `Watching` instance is created for every compiler instance, of which there are 21 (excluding the wrapping MultiCompiler) in my case.

So, 21 synchronous `_go()` invocations, but then within each synchronous `_go()` invocation, we do `this.compiler.applyPluginsAsync()` with a callback. This is a function provided by the Tapable dependency:
```
Tapable.prototype.applyPluginsAsyncSeries = Tapable.prototype.applyPluginsAsync = function applyPluginsAsync(name) {
	var args = Array.prototype.slice.call(arguments, 1);
	var callback = args.pop();
	if(!this._plugins[name] || this._plugins[name].length === 0) return callback();
	var plugins = this._plugins[name];
	var i = 0;
	var _this = this;
	args.push(copyProperties(callback, function next(err) {
		if(err) return callback(err);
		i++;
		if(i >= plugins.length) {
			return callback();
		}
		plugins[i].apply(_this, args);
	}));
	plugins[0].apply(this, args);
};
```
and it looks quite expensive in terms of potential memory footprint, especially since this will all be running asynchronously.

I'm not really sure whether I'm on the right path here, as this is my first time looking at the source code, but I'm tracing my use case through the source code and looking for places where there could potentially be memory leaks.

**Comment by bebraw** [comment]
Yeah, sounds like a leak. It's likely missing cleanup code somewhere.

As a workaround you could try running your configuration through something like [parallel-webpack](https://www.npmjs.com/package/parallel-webpack). This will give you a performance boost as it's able to run in parallel (better core utilization) and it would work around the issue effectively.

Of course if you can track down the leak and send a PR, I'm sure that would be appreciated. 👍

**Comment by helloyou2012** [comment]
@bebraw  `MultiCompiler.js` has plan to support parallel?

**Comment by ayc92** [comment]
@bebraw: Thanks for the suggestion. I'll definitely look into the parallel-webpack package as well as maybe getting a PR up. The parallel-webpack solution is enticing, but I'm actually using [django-webpack](https://github.com/owais/django-webpack-loader), which needs to write the chunk names and their respective bundles to a manifest, and I'm not sure that parallel-webpack handles parallel writes to a file (probably not).

**Comment by bebraw** [comment]
@helloyou2012 I don't know. At least now you have to handle possible parallelization at your end.

@ayc92 Ok, that's harder then. Good luck. 👍

**Comment by sokra** [comment]
I think this just takes so much memory. Keep in mind that node.js also has a memory limit, which can/need to be increased (`--max_old_space_size`).

btw. Watching + MultiCompiler + UglifyJs combined is propably very very slow... What's the usecase for this?

But I would recommend to consider parallel-webpack and integration into your workflow. trivago (@pago) does something similar (as far as I remember 32 compilations).

**Comment by ayc92** [comment]
@sokra: My use case is, I need to conditionally require `jquery-validation` messages based on the locale (see https://github.com/jzaefferer/jquery-validation/tree/master/src/localization), and I'm doing it by using `webpack.DefinePlugin` to set a global variable, that my file then uses like so:
```
if (process.env.LOCALE !== 'en-us') {
    require("jquery-validation/dist/localization/messages_" + process.env.JQUERY_VALIDATION_LOCALE);
}
```
Unfortunately, if I do it this way, I have to generate a bundle for each locale. 

I am not in fact using UglifyJS in my development environment, and also, I run out of memory regardless of whether I'm watching or not. I've also tried increasing node's old space size limit, to no avail.

I'm trying to use the analyse tool at the moment, but it's a bit difficult to grasp.

**Comment by ayc92** [comment]
So, I got it to work without the OOM error. I realized that I didn't need to loop over all languages for all of my entry points; I only needed to do this for the "microsite" entry point. So I fixed that.

In addition, for any shared plugins, I made sure to instantiate them only once. I'm not certain this is an acceptable way of setting up webpack, but it seems to work just fine, and I'd imagine for a large number of configs, it's a considerable savings on memory usage.
